File: end-to-end-applications/typescript/food-ordering/app/src/delivery-app/delivery_manager.ts
/*
 * Copyright (c) 2024 - Restate Software, Inc., Restate GmbH
 *
 * This file is part of the Restate examples,
 * which is released under the MIT license.
 *
 * You can find a copy of the license in the file LICENSE
 * in the root directory of this repository or package or at
 * https://github.com/restatedev/examples/
 */

import * as restate from "@restatedev/restate-sdk";
import * as driverDigitalTwin from "./driver_digital_twin";
import * as driverDeliveryMatcher from "./driver_delivery_matcher";
import * as geo from "./utils/geo";
import {DEMO_REGION, Location, DeliveryInformation, OrderAndPromise, Status} from "./types/types";
import * as orderstatus from "../order-app/order_status_service";

/**
 * Manages the delivery of the order to the customer. Keyed by the order ID (similar to the
 * OrderService and OrderStatusService).
 */
export const service: restate.ServiceApi<typeof router> = { path: "delivery-manager" };

const DELIVERY_INFO = "delivery-info";

export const router = restate.keyedRouter({
    // Called by the OrderService when a new order has been prepared and needs to be delivered.
    start: async (ctx: restate.KeyedContext, deliveryId: string, { order, promise }: OrderAndPromise) => {
        const [restaurantLocation, customerLocation] = await ctx.sideEffect(async () => [geo.randomLocation(), geo.randomLocation()]);

        // Store the delivery information in Restate's state store
        const deliveryInfo: DeliveryInformation = {
            orderId: order.id,
            orderPromise: promise,
            restaurantId: order.restaurantId,
            restaurantLocation,
            customerLocation,
            orderPickedUp: false
        }
        ctx.set(DELIVERY_INFO, deliveryInfo);

        // Acquire a driver
        const driverPromise = ctx.awakeable<string>();
        ctx.send(driverDeliveryMatcher.service).requestDriverForDelivery(DEMO_REGION, { promiseId: driverPromise.id });
        // Wait until the driver pool service has located a driver
        // This awakeable gets resolved either immediately when there is a pending delivery
        // or later, when a new delivery comes in.
        const driverId = await driverPromise.promise;

        // Assign the driver to the job
        await ctx.rpc(driverDigitalTwin.service).assignDeliveryJob(driverId, {
                deliveryId,
                restaurantId: order.restaurantId,
                restaurantLocation: deliveryInfo.restaurantLocation,
                customerLocation: deliveryInfo.customerLocation
            });

        ctx.send(orderstatus.service).setStatus(order.id, Status.WAITING_FOR_DRIVER);
    },

    // called by the DriverService.NotifyDeliveryPickup when the driver has arrived at the restaurant.
    notifyDeliveryPickup: async (ctx: restate.KeyedContext, _deliveryId: string) => {
        const delivery = (await ctx.get<DeliveryInformation>(DELIVERY_INFO))!;
        delivery.orderPickedUp = true;
        ctx.set(DELIVERY_INFO, delivery);

        ctx.send(orderstatus.service).setStatus(delivery.orderId, Status.IN_DELIVERY);
    },

    // Called by the DriverService.NotifyDeliveryDelivered when the driver has delivered the order to the customer.
    notifyDeliveryDelivered: async (ctx: restate.KeyedContext, _deliveryId: string) => {
        const delivery = (await ctx.get<DeliveryInformation>(DELIVERY_INFO))!;
        ctx.clear(DELIVERY_INFO);

        // Notify the OrderService that the delivery has been completed
        ctx.resolveAwakeable(delivery.orderPromise, null);
    },

    // Called by DriverDigitalTwin.HandleDriverLocationUpdateEvent() when the driver moved to new location.
    handleDriverLocationUpdate: async (ctx: restate.KeyedContext, _deliveryId: string, location: Location) => {
        const delivery = (await ctx.get<DeliveryInformation>(DELIVERY_INFO))!;

        // Parse the new location, and calculate the ETA of the delivery to the customer
        const eta = delivery.orderPickedUp
            ? geo.calculateEtaMillis(location, delivery.customerLocation)
            : geo.calculateEtaMillis(location, delivery.restaurantLocation)
            + geo.calculateEtaMillis(delivery.restaurantLocation, delivery.customerLocation);
        ctx.send(orderstatus.service).setETA(delivery.orderId, eta);
    },
});

File: end-to-end-applications/python/food-ordering/app/ordering/delivery_manager.py
from restate import VirtualObject, ObjectContext
from restate.exceptions import TerminalError

from ordering.order_workflow import Order
import ordering.order_workflow as order_workflow
import ordering.driver_matcher as driver_matcher
import ordering.driver_digital_twin as driver_digital_twin
import ordering.order_status as order_status
from ordering.utils import geo
from ordering.types.types import DeliveryInformation, Location

from ordering.types.types import DEMO_REGION

delivery_manager = VirtualObject("delivery-manager")

DELIVERY_INFO = "DELIVERY_INFO"


@delivery_manager.handler()
async def start(ctx: ObjectContext, order: Order):
    restaurant_location, customer_location = await ctx.run("locations", lambda: [
        geo.random_location(),
        geo.random_location()
    ])

    # Store the delivery information in Restate's state store
    delivery_info: DeliveryInformation = {
        "order_id": order["id"],
        "restaurant_id": order["restaurant_id"],
        "restaurant_location": restaurant_location,
        "customer_location": customer_location,
        "order_picked_up": False,
    }
    ctx.set(DELIVERY_INFO, delivery_info)

    # Acquire a driver
    driver_promise_id, driver_promise = ctx.awakeable()

    ctx.object_send(driver_matcher.request_driver_for_delivery, DEMO_REGION, {"promise_id": driver_promise_id})

    # Wait until the driver pool service has located a driver
    driver_id = await driver_promise

    # Assign the driver to the job
    await ctx.object_call(driver_digital_twin.assign_delivery_job, driver_id, {
        "delivery_id": ctx.key(),
        "restaurant_id": order["restaurant_id"],
        "restaurant_location": delivery_info["restaurant_location"],
        "customer_location": delivery_info["customer_location"],
    })

    await ctx.workflow_call(order_workflow.selected_driver, order["id"], arg=None)


@delivery_manager.handler()
async def notify_delivery_pickup(ctx: ObjectContext):
    delivery = await ctx.get(DELIVERY_INFO)
    if delivery is None:
        raise TerminalError("No delivery information found")

    delivery["order_picked_up"] = True
    ctx.set(DELIVERY_INFO, delivery)

    ctx.workflow_send(order_workflow.signal_driver_at_restaurant, delivery["order_id"], arg=None)


@delivery_manager.handler()
async def notify_delivery_delivered(ctx: ObjectContext):
    delivery = await ctx.get(DELIVERY_INFO)
    if delivery is None:
        raise TerminalError("No delivery information found")
    ctx.clear(DELIVERY_INFO)

    ctx.workflow_send(order_workflow.signal_delivery_finished, delivery["order_id"], arg=None)


@delivery_manager.handler("handleDriverLocationUpdate")
async def handle_driver_location_update(ctx: ObjectContext, location: Location):
    delivery = await ctx.get(DELIVERY_INFO)

    if delivery["order_picked_up"]:
        eta = geo.calculate_eta_millis(location, delivery["customer_location"])
    else:
        eta = (geo.calculate_eta_millis(location, delivery["restaurant_location"]) +
               geo.calculate_eta_millis(delivery["restaurant_location"], delivery["customer_location"]))

    ctx.object_send(order_status.set_eta, delivery["order_id"], eta)
